### PR TITLE
storage: multi-worker postgres source refactor

### PR DESCRIPTION
**Pull request by petrosagg**
## Motivation

The original motivation for this work was to optimize the ingestion of postgres sources by using all the available timely workers but in the process many more improvements were realized.

Beyond the better performance characteristics this new implementation improves upon the previous on the following additional dimensions:

### Error Handling
Previously each error could be tagged as definite/indefinite/irrecoverable by attaching a suffix function call like `.err_definite()` and requests to restart the dataflow had to be manually wired up. 

**Transient error handling:** This new work introduces a new operator builder function (`AsyncOperatorBuilder::build_fallible`) which allows using of the standard `?` operator to write fallible operators. This is how all transient errors are handled and they all lead to the dataflow restarting. 

**Definite error handling:** A concrete `DefiniteError` enum type is introduced with one variant per expected failure. There is no "catchall" variant which enforces the necessary discipline around definite errors. `DefiniteError` is the only kind of error that is allowed to end up in a collection.

### Semantics of LSNs

In the code on main there were plenty of places where we were adding or subtracting one to LSNs without justification. I have investigated all the interactions with postgres to investigate how it interprets LSNs (i.e if pg treats them as `<` or `<=`) and we know have the correct conversions when the timely conventions differ from the pg conventions with a comment explaining the behavior.

Additionally, I discovered that we were incorrectly using the `commit_body.end_lsn()` to timestamp transactions instead of the `commit_body.commit_lsn()`, which is always equal to `begin_body.final_lsn()`.

### Separation of concerns

We previously had a very large async event loop with multiple labeled `loop` statements where control flow was hard to follow. In this work I have separated the handling of the replication stream in three distinct components: 

1. Extracting `LogicalReplicationMessages` while handling standby status updates (function `raw_stream`).
2. Extracting a single transaction from a stream containing `LogicalReplicationMessages` (function `extract_transaction`)
3. Alternating between reading the stream and discovering replication lag (top level loop)

Beyond that the snapshot and replication reading are separated into their respective modules and their results are concatenated at the timely level.

### Documentation

The code is now *extensively* documented with explanations and diagrams of all the components. The main `postgres.rs` module contains a high level overview of the dataflow and `snapshot.rs` and `replication.rs` contain detailed descriptions of their mechanics.

### Succinctness

The code is now significantly shorter and (in my opinion) simpler to follow. The diff count on Github doesn't really do it justice due to the hundreds of lines of documentation added.

## Tips for reviewer

First read `postgres.rs`, starting with its module documentation. There is no point looking at the diff for this file, there aren't any commonalities with the previous code.

Then, read `snapshot.rs` starting with its module documentation and finally read `replication.rs` also starting with its module documentation.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/postgres.rs [hunk 1]
```diff
@@ -7,1664 +7,325 @@
...
+//! `ReplicationReader` operators, which are the ones that talk to the external world. Both of
+//! these operators are built with the `AsyncOperatorBuilder::build_fallible` method which allows
+//! transient errors to be propagated upwards with the standard `?` operator without risking
+//! downgrading the capability and producing bogus frontiers.
+//!
+//! The error streams from both of those operators are published to the source status and also
+//! trigger a restart of the dataflow.
+//!
+//! ```text
+//!    ┏━━━━━━━━━━━━━━┓
```
Comment (aalexandrov) [inline src/storage/src/source/postgres.rs:54]: Did you use a tool to create the diagram or do you have a cheat-sheet of UTF border symbols?

File: src/storage/src/source/postgres.rs [hunk 2]
```diff
@@ -7,1664 +7,325 @@
...
-        datums.push(datum);
-    }
-    Ok(())
+/// Errors that can occur when executing a single row query.
+#[derive(Debug, thiserror::Error)]
+enum SingleRowError {
+    /// The action can only be performed by a superuser.
+    #[error("error executing query: {0}")]
+    Other(#[from] tokio_postgres::Error),
+    /// The query returned more than one row
```
Comment (petrosagg) [inline src/storage/src/source/postgres.rs:301]: copy pasta, will fix

File: src/storage/src/source/postgres.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/source/postgres.rs:54]: I used this https://en.wikipedia.org/wiki/Box-drawing_character#Box_Drawing and a lot of will power

File: src/storage/src/source/postgres/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+//! Given that all our ingestion is based on correctly timestamping updates with the LSN they
+//! happened at it is important that we run the `COPY` query at a specific LSN point that is
+//! relatable with the LSN numbers we receive from the replication stream. Such point does not
+//! necessarily exist for a normal SQL transaction. To achieve this we must force postgres to
+//! produce a consistent point and let us know of the LSN number of that by creating a replication
+//! slot as the first statement in a transaction.
+//!
+//! This is a temporary dummy slot that is only used to put our snapshot transaction on a
+//! consistent LSN point. Unfortunately no lighterweight method exists for doing this. See this
+//! [postgres thread] for more details.
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:29]: We should mention why we don't just create a real replication slot and use that in the `replication` module, instead opting for this temporary one! Also, thank you for linking to the email thread!

File: src/storage/src/source/postgres/snapshot.rs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+//!
+//! [postgres thread]: https://www.postgresql.org/message-id/flat/CAMN0T-vzzNy6TV1Jvh4xzNQdAvCLBQK_kh6_U7kAXgGU3ZFg-Q%40mail.gmail.com
+//!
+//! ## Reusing the consistent point among all workers
+//!
+//! Creating replication slots is potentially expensive so the code makes is such that all workers
+//! cooperate and reuse one consistent snapshot among them. In order to do so we make use the
+//! "export transaction" feature of postgres. This feature allows one SQL session to create an
+//! identifier for the transaction (a string identifier) it is currently in which can be used by
+//! other sessions to enter the same "snapshot".
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:39]:
```suggestion
//! Creating replication slots is potentially expensive so the code makes is such that all workers
//! cooperate and reuse one consistent snapshot among them. In order to do so we make use the
//! "export transaction" feature of postgres. This feature allows one SQL session to create an
//! identifier for the transaction (a string identifier) it is currently in, which can be used by
//! other sessions to enter the same "snapshot".
```

File: src/storage/src/source/postgres/snapshot.rs [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+//!
+//! We accompplish this by picking one worker at random to function as the transaction leader. The
+//! transaction leader is responsible for starting a SQL session, creating a temporary replication
+//! slot in a transaction, exporting the transaction id, and broadcasting it the transaction
+//! information to all other workers via a broadcasted feedback edge.
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:44]:
```suggestion
//! slot in a transaction, exporting the transaction id, and broadcasting the transaction
//! information to all other workers via a broadcasted feedback edge.
```

File: src/storage/src/source/postgres/snapshot.rs [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+//! ## Coordinated transaction COMMIT
+//!
+//! When follower workers are done with snapshotting they commit their transaction, close their
+//! session, and then drop their snapshot feedback capability. When the leader worker is done with
+//! snapshotting it drops its snapshot feedback capability and waits until it observes the
+//! snapshot input advancing to the empty frontier. This allows the leader to COMMIT its
+//! transaction last, which is the transaction that exported the snapshot.
+//!
+//! It's unclear if this is strictly necessary, but having the frontiers made it easy enough that I
+//! added the synchronization.
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:62]: very nice!

File: src/storage/src/source/postgres/snapshot.rs [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+            let id = config.id;
+            let worker_id = config.worker_id;
+
+            let [data_cap, rewind_cap, snapshot_cap]: &mut [_; 3] = caps.try_into().unwrap();
+            let data_cap = data_cap.as_mut().unwrap();
+            trace!(%id, "timely-{worker_id} initializing table reader {}", resume_upper.pretty());
+
+            if resume_upper.into_option() != Some(MzOffset::from(0)) {
+                return Ok(());
+            }
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:199]:
```suggestion
            // We have already snapshotted, so the replication dataflow will take over
            if resume_upper.into_option() != Some(MzOffset::from(0)) {
                return Ok(());
            }
```

Comment (guswynn) [reply src/storage/src/source/postgres/snapshot.rs:199]: something like that ^

File: src/storage/src/source/postgres/snapshot.rs [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+                let snapshot_info = export_snapshot(&client).await?;
+                trace!(%id, "timely-{worker_id} exporting snapshot info {snapshot_info:?}");
+                let cap = snapshot_cap.as_ref().unwrap();
+                snapshot_handle.give(cap, snapshot_info).await;
+            }
+
+            let (snapshot, snapshot_lsn) = match snapshot_input.next_mut().await {
+                Some(AsyncEvent::Data(_, data)) => data.pop().expect("snapshot sent above"),
+                _ => unreachable!(),
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:222]:
```suggestion
                _ => unreachable!("Data events always proceed Progress ones"),
```

File: src/storage/src/source/postgres/snapshot.rs [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+                                continue;
+                            }
+                        }
+                    }
+                    None => {
+                        let err = Err(DefiniteError::TableDropped);
+                        raw_handle.give(data_cap, ((oid, err), MzOffset::minimum(), 1)).await;
+                        continue;
+                    }
+                };
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:263]: is it possible to do this verification in a function?

File: src/storage/src/source/postgres/snapshot.rs [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+                Err(anyhow!("recoverable errors should crash the process during snapshots"))
+            });
+
+            // The exporting worker should wait for all the other workers to commit before dropping
+            // its client since this is what holds the exported transaction alive.
+            if is_snapshot_leader {
+                trace!(%id, "timely-{worker_id} waiting for all workers to finish");
+                *snapshot_cap = None;
+                while snapshot_input.next().await.is_some() {}
+                trace!(%id, "timely-{worker_id} comitting COPY transaction");
```
Comment (guswynn) [inline src/storage/src/source/postgres/snapshot.rs:290]:
```suggestion
                trace!(%id, "timely-{worker_id} (leader) comitting COPY transaction");
```

File: src/storage/src/source/postgres/replication.rs [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+
+            if !config.responsible_for("slot") {
+                return Ok(());
+            }
+
+            let Some(resume_lsn) = resume_upper.into_option() else {
+                return Ok(());
+            };
+            data_cap.downgrade(&resume_lsn);
+            upper_cap.downgrade(&resume_lsn);
```
Comment (guswynn) [inline src/storage/src/source/postgres/replication.rs:148]: god I love source-timestamp-timely

File: src/storage/src/source/postgres/replication.rs [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+                    }
+                }
+            }
+
+            let mut committed_uppers = pin!(committed_uppers);
+
+            loop {
+                // TODO(guswynn|petrosagg): fix the underlying bug that prevents client re-use when
+                // exiting the CopyBoth mode, so we don't need to re-create clients in this function.
+                client = connection_config.connect("advance_upper").await?;
```
Comment (guswynn) [inline src/storage/src/source/postgres/replication.rs:177]: is this string just for an error message?

File: src/storage/src/source/postgres/replication.rs [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+
+                if let Err(err) = ensure_publication_exists(&client, &connection.publication).await? {
+                    // If the publication gets deleted there is nothing else to do
```
Comment (guswynn) [inline src/storage/src/source/postgres/replication.rs:180]:
```suggestion
                    // If the publication gets deleted there is nothing else to do. These errors are not retractable.
```

File: src/storage/src/source/postgres/replication.rs [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+                    for &oid in table_info.keys() {
+                        let update = ((oid, Err(err.clone())), *data_cap.time(), 1);
+                        data_output.give(data_cap, update).await;
+                    }
+                    return Ok(());
+                }
+
+                // We will peek the slot first to eagerly advance the frontier in case the stream
+                // is currently empty. This unblocks downstream consumers immediately instead of
+                // making them wait for TX_TIMEOUT to trigger.
```
Comment (guswynn) [inline src/storage/src/source/postgres/replication.rs:190]: can you clarify this? what is a "downstream consumer"?

File: src/storage/src/source/postgres/replication.rs [hunk 15]
```diff
@@ -0,0 +1,556 @@
...
+                    for req in data.drain(..) {
+                        assert!(min_resume_lsn <= req.snapshot_lsn, "slot compacted past snapshot point. snapshot_lsn={} min_resume_lsn={min_resume_lsn}", req.snapshot_lsn);
+                        rewinds.insert(req.oid, (cap.clone(), req));
+                    }
+                }
+            }
+
+            let mut committed_uppers = pin!(committed_uppers);
+
+            loop {
```
Comment (guswynn) [inline src/storage/src/source/postgres/replication.rs:174]: can you add a brief comment about what this loop retries?

**Review by guswynn — COMMENTED** [review]
This looks amazing the core design seems sound to me! I always wanted to add a task to https://github.com/MaterializeInc/materialize/issues/17598 that was "actually think about what the pg implementation should look like", which this pr definitely does!

I have not dug into the exact capability handling and various subtle points here, because I have 2 overarching questions that I'll need answer first:

- `pg_export_snapshot` is awesome! Does it use additional disk space, and does this space leak if the snapshot fails midway? Even more broadly, does the snapshot its exporting inherit the temporary nature of the snapshot?
- The most difficult part of this is the rewind logic.
  - When we first start up, we start a snapshot dataflow, which produces a snapshot at some `snapshot_lsn`, produces the snapshot stream, and issues `RewindRequest`'s at `snapshot_lsn`. The replication stream starts up, usually with the same lsn (which it determines from the `RewindRequest`'s?. When do we actually ever need to do retractions? If we restart before committing the first snapshot?

File: src/storage/src/source/postgres/replication.rs [hunk 16]
```diff
@@ -0,0 +1,556 @@
...
+            None => Err(anyhow!("stream ended prematurely"))?,
+        };
+        metrics.transactions.inc();
+        metrics.lsn.set(commit_lsn.into());
+        while let Some(event) = stream.try_next().await? {
+            metrics.total.inc();
+            match event {
+                Insert(body) if errored_tables.contains(&body.rel_id()) => continue,
+                Update(body) if errored_tables.contains(&body.rel_id()) => continue,
+                Delete(body) if errored_tables.contains(&body.rel_id()) => continue,
```
Comment (sploiselle) [inline src/storage/src/source/postgres/replication.rs:384]: Should also have one of these checks for `Relation`

File: src/storage/src/source/postgres/replication.rs [hunk 17]
```diff
@@ -0,0 +1,556 @@
...
+                            None => {
+                                errored_tables.insert(rel_id);
+                                yield (rel_id, Err(DefiniteError::TableDropped), 1);
+                            }
+                        }
+                    }
+                }
+                Truncate(body) => {
+                    for &rel_id in body.rel_ids() {
+                        if errored_tables.contains(&rel_id) {
```
Comment (sploiselle) [inline src/storage/src/source/postgres/replication.rs:454]: nit: this is a little more succinct as:

```rust
if errored_tables.insert(rel_id) {
    yield (rel_id, Err(DefiniteError::TableTruncated), 1);
}
```

File: src/storage/src/source/postgres/snapshot.rs [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+
+    let mut snapshot_input = builder.new_input(&snapshot_feedback, Pipeline);
+    let (mut raw_handle, raw_data) = builder.new_output_connection(vec![Antichain::new()]);
+    let (mut rewinds_handle, rewinds) = builder.new_output_connection(vec![Antichain::new()]);
+
+    // The export id must be sent to all workes, so we broadcast the feedback connection
+    let (mut snapshot_handle, snapshot) = builder.new_output_connection(vec![Antichain::new()]);
+    snapshot.broadcast().connect_loop(feedback_handle);
+
+    let is_snapshot_leader = config.responsible_for("snapshot_leader");
```
Comment (sploiselle) [inline src/storage/src/source/postgres/snapshot.rs:183]: It seems odd to me that we choose the same random worker for all PG sources. The idea is just that we improve this over choosing `0`?

File: src/storage/src/source/postgres/snapshot.rs [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+                trace!(%id, "timely-{worker_id} exporting snapshot info {snapshot_info:?}");
+                let cap = snapshot_cap.as_ref().unwrap();
+                snapshot_handle.give(cap, snapshot_info).await;
+            }
+
+            let (snapshot, snapshot_lsn) = match snapshot_input.next_mut().await {
+                Some(AsyncEvent::Data(_, data)) => data.pop().expect("snapshot sent above"),
+                _ => unreachable!(),
+            };
+            if !is_snapshot_leader {
```
Comment (sploiselle) [inline src/storage/src/source/postgres/snapshot.rs:224]: Can we add a comment like:

> Snapshot leader is already in identified transaction but all other workers need to enter it.

File: src/storage/src/source/postgres/snapshot.rs (hunk 9 above)
Comment (sploiselle) [reply src/storage/src/source/postgres/snapshot.rs:263]: plus 1

File: src/storage/src/source/postgres/snapshot.rs [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+//!
+//! It's unclear if this is strictly necessary, but having the frontiers made it easy enough that I
+//! added the synchronization.
+//!
+//! ## Snapshot rewinding
+//!
+//! Ingestion dataflows must produce definite data, including the snapshot. What this means
+//! practically is that whenever we deem it necessary to snapshot a table we must do so at the same
+//! LSN. However, the method for running a transaction described above doesn't let us choose the
+//! LSN, it could be an LSN in the future chosen by postres while it creates the temporary
```
Comment (sploiselle) [inline src/storage/src/source/postgres/snapshot.rs:69]:
```suggestion
//! LSN, it could be an LSN in the future chosen by PG while it creates the temporary
```

File: src/storage/src/source/postgres/snapshot.rs [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+//!```text
+//! sum(update: t <= t_snapshot) = sum(update: t <= t_slot) + sum(update: t_slot <= t <= t_snapshot)
+//!                                         |
+//!                                         V
+//! sum(update: t <= t_slot) = sum(update: t <= snapshot) - sum(update: t_slot <= t <= t_snapshot)
+//! ```
+//!
+//! Therefore, if we manage to recover the `sum(update: t_slot <= t <= t_snapshot)` term we will be
+//! able to "rewind" the snapshot we obtained at `t_snapshot` to `t_slot` by emitting all updates
+//! that happen inbetween these two points with their diffs negated.
```
Comment (sploiselle) [inline src/storage/src/source/postgres/snapshot.rs:89]:
```suggestion
//! that happen between these two points with their diffs negated.
```

File: src/timely-util/src/builder_async.rs [hunk 22]
```diff
@@ -524,6 +524,68 @@ impl<G: Scope> OperatorBuilder<G> {
         button
     }
 
+    /// Creates a fallible operator implementation from supplied logic constructor. If the `Future`
+    /// resolves to an error it will emitted in the returned error stream and then the operator
```
Comment (sploiselle) [inline src/timely-util/src/builder_async.rs:528]:
```suggestion
    /// resolves to an error it will be emitted in the returned error stream and then the operator
```

**Review by sploiselle — APPROVED** [review]
I have the smallest nits. Gus' asks for add'l comments are good, as well so I'd like to see those before we merge.

Extracting schema checks to its own function seems good but if it's not trivial I wouldn't sweat it too much.

**Comment by philip-stoev** [comment]
Nightly is clean, that is, failures are unrelated.

I have started a the pg-cdc step of the Release Qualification build, so  once it is complete feel free to merge . I have no outstanding ideas around furher testing pg-cdc failures, so the existing tests are what could be brought to bear against this PR.

**Comment by philip-stoev** [comment]
If you can rebase once again, you can run the feature benchmark, just in case.

**Comment by philip-stoev** [comment]
Item No  1. Unfortunately the long pg-cdc release qualification zippy test shows a failure that has not been seen before -- the storaged container failed to start . Unfortunately, Buildkite did not capture any logs so it could as well be a docker problem, we have had a lot of those recently albeit with other manifestations. I am now rerunning the job.

File: src/storage/src/source/postgres.rs [hunk 23]
```diff
@@ -7,1664 +7,341 @@
...
+            Antichain::from_iter(config.source_resume_upper.iter().map(MzOffset::decode_row));
+
+        // Collect the tables that we will be ingesting.
+        let mut table_info = BTreeMap::new();
+        let mut subsource_outputs = vec![];
+        for (i, desc) in self.publication_details.tables.iter().enumerate() {
+            // Index zero maps to the main source
+            let output_index = i + 1;
+            // We maintain descriptions for all tables in the publication, but only casts for those
+            // we aim to use (and have validated that their types). This prevents us from creating
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/postgres.rs:141]: The publication might contain more tables than the user has selected to ingest (via a restricted FOR TABLES <..>). The tables that are to be ingested will be present in the table_casts map and so we can filter the publication tables based on whether or not we have casts for it.

I replaced the comment in the code with this ^ text

File: src/storage/src/source/postgres.rs [hunk 24]
```diff
@@ -7,1664 +7,341 @@
...
-            let offset_commit_metrics = source_metrics.offset_commit_metrics();
-
-            let mut reader = PostgresSourceReader {
-                receiver_stream: dataflow_rx,
-                last_lsn: start_offset.offset.into(),
-                data_capability,
-                upper_capability,
+        let health = snapshot_err.concat(&repl_err).flat_map(move |err| {
+            let update = HealthStatus::StalledWithError {
+                error: format!("{:#}", err),
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/postgres.rs:181]: I am using `anyhow::Error` and `{:#}` is how you get that

File: src/storage/src/source/postgres.rs [hunk 25]
```diff
@@ -7,1664 +7,341 @@
...
-                    // the select! loop so the future stays on the stack and never gets cancelled
-                    // until the end of the function.
-                    _ = offset_commit_loop.as_mut() => {},
-                }
-            }
+            // But we still want to report the transient error for all subsources
+            statuses.extend(subsource_outputs.iter().map(|index| {
+                let status = HealthStatusUpdate {
+                    update: update.clone(),
+                    should_halt: false,
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/postgres.rs:195]: No, we assert in the health operator that only output 0 can request a halt. This was surprising to me too but I don't want to fold a health operator refactor into this PR

File: src/timely-util/src/builder_async.rs [hunk 26]
```diff
@@ -524,6 +524,68 @@ impl<G: Scope> OperatorBuilder<G> {
...
+    ///
+    ///     // Explicitly dropping the capability:
+    ///     // Simply running `drop(cap)` will only drop the reference and not the capability itself!
+    ///     *cap = None;
+    ///
+    ///     // !! BIG WARNING !!:
+    ///     // It is tempting to `take` the capability out of the option for convenience. This will
+    ///     // move the capability into the future state, tying its lifetime to it, which will get
+    ///     // dropped when an error is hit, causing incorrect progress statements.
+    ///     let cap = cap.take().unwrap(); // DO NOT DO THIS
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/timely-util/src/builder_async.rs:559]: Because the logic needs to be able to drop the capability. if you did `foo = None` on a `Option<&mut Cap>` it would just drop the mut ref, which is a no-op.

File: src/storage/src/source/postgres/snapshot.rs (hunk 18 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/snapshot.rs:183]: We choose a different random worker for each pg source because the `responsible_for` function factors in the hash of the global id together with the partition

File: src/timely-util/src/builder_async.rs [hunk 27]
```diff
@@ -524,6 +524,68 @@ impl<G: Scope> OperatorBuilder<G> {
...
+    /// }));
+    /// ```
+    pub fn build_fallible<E: 'static, F>(
+        mut self,
+        constructor: F,
+    ) -> (ShutdownButton<()>, StreamCore<G, Vec<Rc<E>>>)
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/timely-util/src/builder_async.rs:565]: This function doesn't care what the `E` type is. Whoever is using it will either have a concrete `Rc<Foo> whree Foo: Error` type (which is the case here) or can add additional bounds on `E` so that `Error` methods can be called.

It's annoying to not have the `Rc` here because timely *really* insists that all types are clonable even if you never actually clone them and many error types *really* insist that they can't be cloned so you end up doing a whole lot of wrapping/unwrapping.

File: src/storage/src/source/postgres.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/postgres.rs:181]: Actually I switched to a custom `TransientError` enum and changed this to `display_with_causes().to_string()`

File: src/storage/src/source/postgres/snapshot.rs (hunk 8 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/snapshot.rs:222]: Good catch! I removed the `unreachable` because I don't want to depend on how the handle produces events

File: src/storage/src/source/postgres/snapshot.rs (hunk 18 above)
Comment (sploiselle) [reply src/storage/src/source/postgres/snapshot.rs:183]: Very clever, ty for the ponter

File: src/storage/src/source/postgres/snapshot.rs (hunk 9 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/snapshot.rs:263]: Yes!

File: src/storage/src/source/postgres/replication.rs (hunk 14 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/replication.rs:190]: I clarified the comment. Downstream consumer is any operator that consumes the output of this operator (e.g reclocking and persit_sink). This is avoiding holding back the capabilty

**Comment by petrosagg** [comment]
@guswynn @sploiselle I addressed all your review comments and rebased on top of main to fix the conflicts (cc @philip-stoev)


> `pg_export_snapshot` is awesome! Does it use additional disk space, and does this space leak if the snapshot fails midway? Even more broadly, does the snapshot its exporting inherit the temporary nature of the snapshot?


You can ready the details here https://www.postgresql.org/docs/9.4/functions-admin.html#FUNCTIONS-SNAPSHOT-SYNCHRONIZATION but no it doesn't use extra space and yes it is temporary. It just gives a name to an in-memory data structure that would exist anyway.

> - The most difficult part of this is the rewind logic.
>   - When we first start up, we start a snapshot dataflow, which produces a snapshot at some `snapshot_lsn`, produces the snapshot stream, and issues `RewindRequest`'s at `snapshot_lsn`. The replication stream starts up, usually with the same lsn (which it determines from the `RewindRequest`'s?. When do we actually ever need to do retractions? If we restart before committing the first snapshot?

Replication and snapshot always start at different LSNs because they create two separate slots (a temporary and a real one) and creating a slot advances the WAL so there will be a small difference always.

> which produces a snapshot at some `snapshot_lsn`, 

We take a snapshot at `snapshot_lsn` but we produce it at LSN 0, and we do the same for any rewound correction. This is what causes the correction updates to be consolidated with the snapshot updates and end up with a snapshot at `slot_lsn` emitted at LSN 0.

 Is the "snapshot rewind" section of the `snapshot` module documentation unclear? If so it's better to record any clarifications there.

File: src/storage/src/source/postgres/snapshot.rs (hunk 7 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/snapshot.rs:204]: added your suggestion, thanks

File: src/storage/src/source/postgres/replication.rs (hunk 15 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/replication.rs:182]: yes, added!

File: src/storage/src/source/postgres/replication.rs (hunk 16 above)
Comment (petrosagg) [reply src/storage/src/source/postgres/replication.rs:391]: Good catch! fixed

File: src/storage/src/source/postgres.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs]: thanks!

File: src/storage/src/source/postgres.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs]: nice!

File: src/storage/src/source/postgres.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:197]: sounds good!

File: src/timely-util/src/builder_async.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/timely-util/src/builder_async.rs:566]: ah i see :(

File: src/timely-util/src/builder_async.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/timely-util/src/builder_async.rs:560]: ah yes!

**Comment by guswynn** [comment]
> Replication and snapshot always start at different LSNs because they create two separate slots (a temporary and a real one) and creating a slot advances the WAL so there will be a small difference always.

Thank you for this clarification!

> Is the "snapshot rewind" section of the snapshot module documentation unclear? If so it's better to record any clarifications there.

I think a bit more clarification could be added there, especially what you mentioned about the slots being different, and producing the snapshot at lsn 0. Also, I think ALL my review is resolved, except for one final question: can we document the _reason why we produce the snapshot and replication stream with different slots_? The extent of my understanding is that it makes correctness in the face of restarts much easier to reason about, but its unclear to me if there is a more fundamental reason than this!
